### PR TITLE
timeouts and exceptions

### PR DESCRIPTION
**Pull request by coot**
- io-sim: fixed timeouts in presence of exceptions
- io-sim: code style
- io-sim: track reason why a thread is blocked
- io-sim: refactored timeout implementation
- io-sim: improved TimeoutException Show instance

### CODE REVIEW

**Comment by coot** [comment]
The following tests have passed:
* [ ] `network-mux:test`
* [ ] `ouroboros-network-framework:test`
* [ ] `ouroboros-network:test` (including nightly ones)

File: io-sim/src/Control/Monad/IOSim/Internal.hs [hunk 1]
```diff
@@ -129,7 +131,7 @@ data TimerCompletionInfo s =
      | TimerThreadDelay !ThreadId !TimeoutId
      -- ^ `threadDelay` timer run by `ThreadId` which was assigned the given
      -- `TimeoutId` (only used to report in a trace).
-     | TimerTimeout !ThreadId !TimeoutId !(STRef s IsLocked)
+     | TimerTimeout !ThreadId !TimeoutId !(TMVar (IOSim s) ThreadId)
```
Comment (bolt12) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:134]: Why changing to a TMVar?

File: io-sim/src/Control/Monad/IOSim/Internal.hs [hunk 2]
```diff
@@ -244,47 +246,27 @@ schedule !thread@Thread{
...
+            threadAction :: IOSim s ()
+            threadAction =
+              if v then unsafeUnregisterTimeout tmid
+                   else atomically (takeTMVar lock) >>= killThread
+
+            thread' =
+              thread { threadControl =
+                        ThreadControl (case threadAction of
+                                        IOSim k' -> k' (\() -> k (Just x)))
+                                      ctl'
```
Comment (bolt12) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:267]: Why do you need pattern matching here?

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs [hunk 3]
```diff
@@ -1623,6 +1598,15 @@ execWriteTVar :: TVar s a -> a -> ST s ()
 execWriteTVar TVar{tvarCurrent} = writeSTRef tvarCurrent
 {-# INLINE execWriteTVar #-}
 
+execTryPutTMVar :: TVar s (Maybe a) -> a -> ST s Bool
+execTryPutTMVar var a = do
```
Comment (bolt12) [inline io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1602]: This function is called `execTryPutTMVar` but receives a `TVar`

**Review by bolt12 — APPROVED** [review]

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 2 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:267]: We need `SimA` not `IOSim`.

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 1 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:134]: The primary reason was that I wanted to write thread actions using `IOSim` not `SimA`.  It turned out that I had to use the `execReadTVar` but in a limited way.

We actually should use an `MVar`, but it doesn't matter in our use case.  We have only one thread that runs `takeMVar`, so we don't need fairness.  We have two threads that try to write to it, but it's guaranteed that one of them receives the `TMVar` empty, so fairness is not a problem here either.

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs (hunk 3 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1602]: The alternative is:
```
execTryPutTMVar :: TMVar (IOSim s) a -> a -> ST s Bool
```

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 1 above)
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:134]: Okay got it! I don't mind the change!

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 2 above)
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:267]: What do you mean with that? What about the other places where we don't do pattern match? It is not clear to me why in this case we need to do this

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs (hunk 3 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSimPOR/Internal.hs]: I changed to `TMVar`s.

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 2 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:267]: because in other palces we use `SimA` to write thee syntax tree.

Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:267]: Oh nevermind forgot that threadAction existed and it's of type IOSim.

**Pull request merged**